### PR TITLE
Add a profiler

### PR DESCRIPTION
**Pull request by zheng-da**
*Description of changes:*
We need a profiler to understand the runtime performance. This PR enables profiling in training.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by isratnisa — COMMENTED** [review]
Can we throw an error or warning when `len(self._checkpoints)` < 2?

File: python/graphstorm/config/argument.py [hunk 1]
```diff
@@ -227,6 +227,14 @@ def save_perf_results_path(self):
             return self._save_perf_results_path
         return None
 
+    @property
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:230]: You also need to update docs.source.configuration

File: python/graphstorm/run/gsgnn_lp/gsgnn_lp.py [hunk 2]
```diff
@@ -40,11 +40,13 @@
 from graphstorm.eval import GSgnnMrrLPEvaluator
 from graphstorm.model.utils import save_embeddings
 from graphstorm.model import do_full_graph_inference
+from graphstorm.utils import rt_profiler
 
 def main(args):
     config = GSConfig(args)
 
     gs.initialize(ip_config=config.ip_config, backend=config.backend)
+    rt_profiler.init(config.profile_path, rank=gs.get_rank())
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_lp/gsgnn_lp.py:49]: Can we turn tr_profiler off?

File: python/graphstorm/trainer/lp_trainer.py [hunk 3]
```diff
@@ -121,34 +122,43 @@ def fit(self, train_loader, num_epochs,
                 self._model.unfreeze_input_encoder()
             # TODO(xiangsx) Support unfreezing gnn encoder and decoder
 
+            rt_profiler.record('train_init')
             for i, (input_nodes, pos_graph, neg_graph, blocks) in enumerate(train_loader):
+                rt_profiler.record('train_sample')
```
Comment (classicsong) [inline python/graphstorm/trainer/lp_trainer.py:127]: If we do not use th.distributed.sync, the profiling result may be less accurate due to load imbalance.

File: python/graphstorm/run/gsgnn_lp/gsgnn_lp.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/run/gsgnn_lp/gsgnn_lp.py:49]: yes, if `profile_path` is None, `rt_profiler` is disabled.

File: python/graphstorm/trainer/lp_trainer.py (hunk 3 above)
Comment (zheng-da) [reply python/graphstorm/trainer/lp_trainer.py:127]: the analyzer tries to compute the variance in each step as well.

Comment (zheng-da) [reply python/graphstorm/trainer/lp_trainer.py:127]: we cannot put barriers in each step.

File: python/graphstorm/config/argument.py (hunk 1 above)
Comment (zheng-da) [reply python/graphstorm/config/argument.py:230]: what do you mean?

File: python/graphstorm/trainer/lp_trainer.py (hunk 3 above)
Comment (isratnisa) [reply python/graphstorm/trainer/lp_trainer.py:127]: Barrier can be inside the profiler.

Comment (zheng-da) [reply python/graphstorm/trainer/lp_trainer.py:127]: done

File: tests/end2end-tests/graphstorm-ec/test.sh [hunk 4]
```diff
@@ -32,6 +32,11 @@ python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_s
 
 error_and_exit $?
 
+echo "**************dataset: Test edge classification, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, inference: mini-batch, with profiling"
+python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ec_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_ec.yaml --part-config /data/movielen_100k_ec_1p_4t/movie-lens-100k.json --num-epochs 1 --profile-path /tmp/ML_ec_profile
+
+error_and_exit $?
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/test.sh:38]: Can you add a check to check the existence of xxx.npy?

File: python/graphstorm/utils.py [hunk 5]
```diff
@@ -213,4 +214,83 @@ def check(self, name):
...
+            for name, runtimes in self._runtime.items():
+                print(name, sum(runtimes) / len(runtimes), "seconds")
+
+    def save_profile(self):
+        """ Save the profiling result to a file.
+        """
+        if self._profile_path is not None:
+            for name in self._runtime:
+                self._runtime[name] = np.array(self._runtime[name])
+            profile_path = os.path.join(self._profile_path, f"{self._rank}.npy")
```
Comment (classicsong) [inline python/graphstorm/utils.py:291]: Can we save a json object so that it is more readable?

File: tests/end2end-tests/graphstorm-lp/test.sh [hunk 6]
```diff
@@ -32,6 +32,11 @@ python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scrip
 
 error_and_exit $?
 
+echo "**************dataset: Movielens, RGCN layer 1, node feat: fixed HF BERT, BERT nodes: movie, inference: mini-batch, negative_sampler: joint, exclude_training_targets: false, with profiling"
+python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scripts/gsgnn_lp --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_lp_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_lp.yaml --num-epochs 1 --eval-frequency 300 --profile-path /tmp/ML_lp_profile
+
+error_and_exit $?
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-lp/test.sh:38]: Same here, please check the profile output exists.

File: tests/end2end-tests/graphstorm-nc/test.sh [hunk 7]
```diff
@@ -32,6 +32,11 @@ python3 -m graphstorm.run.gs_node_classification --workspace $GS_HOME/training_s
 
 error_and_exit $?
 
+echo "**************dataset: MovieLens, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, inference: mini-batch, with profiling"
+python3 -m graphstorm.run.gs_node_classification --workspace $GS_HOME/training_scripts/gsgnn_np/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_nc.yaml --profile-path /tmp/ML_np_profile
+
+error_and_exit $?
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-nc/test.sh:38]: same here.

**Review by classicsong — APPROVED** [review]

**Pull request merged**